src/modules/mc_pos_control/PositionControl/PositionControl.cpp

```diff
@@ -0,0 +1,241 @@
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+/**
+ * @file PositionControl.cpp
+ */
+
+#include "PositionControl.hpp"
+#include <float.h>
+#include <mathlib/mathlib.h>
+#include <ControlMath.hpp>
+#include <px4_defines.h>
+
+using namespace matrix;
+
+void PositionControl::setVelocityGains(const Vector3f &P, const Vector3f &I, const Vector3f &D)
+{
+	_gain_vel_p = P;
+	_gain_vel_i = I;
+	_gain_vel_d = D;
+}
+
+void PositionControl::setVelocityLimits(const float vel_horizontal, const float vel_up, const float vel_down)
+{
+	_lim_vel_horizontal = vel_horizontal;
+	_lim_vel_up = vel_up;
+	_lim_vel_down = vel_down;
+}
+
+void PositionControl::setThrustLimits(const float min, const float max)
+{
+	// make sure the thrust has a tiny minimal length to infer the direction
+	_lim_thr_min = math::max(min, -10e-4f);
+	_lim_thr_max = max;
+}
+
+void PositionControl::setState(const PositionControlStates &states)
+{
+	_pos = states.position;
+	_vel = states.velocity;
+	_yaw = states.yaw;
+	_vel_dot = states.acceleration;
+}
+
+void PositionControl::setInputSetpoint(const vehicle_local_position_setpoint_s &setpoint)
+{
+	_pos_sp = Vector3f(setpoint.x, setpoint.y, setpoint.z);
+	_vel_sp = Vector3f(setpoint.vx, setpoint.vy, setpoint.vz);
+	_acc_sp = Vector3f(setpoint.acceleration);
+	_thr_sp = Vector3f(setpoint.thrust);
+	_yaw_sp = setpoint.yaw;
+	_yawspeed_sp = setpoint.yawspeed;
+}
+
+void PositionControl::setConstraints(const vehicle_constraints_s &constraints)
+{
+	_constraints = constraints;
+
+	// For safety check if adjustable constraints are below global constraints. If they are not stricter than global
+	// constraints, then just use global constraints for the limits.
+	if (!PX4_ISFINITE(constraints.tilt) || (constraints.tilt > _lim_tilt)) {
+		_constraints.tilt = _lim_tilt;
+	}
+
+	if (!PX4_ISFINITE(constraints.speed_up) || (constraints.speed_up > _lim_vel_up)) {
+		_constraints.speed_up = _lim_vel_up;
+	}
+
+	if (!PX4_ISFINITE(constraints.speed_down) || (constraints.speed_down > _lim_vel_down)) {
+		_constraints.speed_down = _lim_vel_down;
+	}
+
+	if (!PX4_ISFINITE(constraints.speed_xy) || (constraints.speed_xy > _lim_vel_horizontal)) {
+		_constraints.speed_xy = _lim_vel_horizontal;
+	}
+}
+
+bool PositionControl::update(const float dt)
+{
+	_positionControl();
+	_velocityControl(dt);
+
+	_yawspeed_sp = PX4_ISFINITE(_yawspeed_sp) ? _yawspeed_sp : 0.f;
+	_yaw_sp = PX4_ISFINITE(_yaw_sp) ? _yaw_sp : _yaw;
+
+	const bool valid_acceleration_setpoint = PX4_ISFINITE(_acc_sp(0))
+			&& PX4_ISFINITE(_acc_sp(1))
+			&& PX4_ISFINITE(_acc_sp(2));
+	return valid_acceleration_setpoint;
+}
+
+void PositionControl::_positionControl()
+{
+	Vector3f vel_sp_position = (_pos_sp - _pos).emult(_gain_pos_p);
+	_addIfNotNanVector(_vel_sp, vel_sp_position);
+	_addIfNotNanVector(vel_sp_position, Vector3f());
```

Comment:
moving into a wrapping function `_setZeroIfNanVector(vel_sp_position)` that contains `_addIfNotNanVector(vel_sp_position, Vector3f());` 

also, maybe invert the two calls such that we clearly see that `_vel_sp` is set to zero if is is NAN and `vel_sp_position` is also NAN

Reply:
> moving into a wrapping function `_setZeroIfNanVector`

Very good idea, makes it more clear. Fixed in https://github.com/PX4/Firmware/pull/12072/commits/ede46989512854320f290e1ffaf492d244324c0e

> also, maybe invert the two calls such that we clearly see that _vel_sp is set to zero if is is NAN and vel_sp_position is also NAN

They cannot be reordered without breaking the functionality because if `vel_sp_position` and `_vel_sp` are both `NAN` there's no velocity setpoint to execute. It's not replaced with zero in that case which allows executing an acceleration setpoint exclusively. If there is no acceleration setpoint the `PositionControl` class will report an error https://github.com/PX4/Firmware/blob/ede46989512854320f290e1ffaf492d244324c0e/src/modules/mc_pos_control/PositionControl/PositionControl.cpp#L114-L117.